### PR TITLE
Add support for OpenPS2Loader VirtualMemoryCards

### PR DESCRIPTION
**Pull request by KrossX**
As requested on Issue #2959, this adds some basic support to OPL VMC files (.bin) that use 512 bytes pages. Normal pages are 528 bytes long, those extra 16 bytes contain 12 bytes of ECC data.

[PS2 Memory Card File System](https://www.psdevwiki.com/ps3/PS2_Savedata#Memory_card_PS2_.28original.29)

On load, the OPL card is converted to a new temporal file in PCSX2's RAW format and ECC is calculated per page. On close, the reverse is done overwriting the original file and deleting the temporal one.

### CODE REVIEW

File: pcsx2/gui/MemoryCardFile.cpp [hunk 1]
```diff
@@ -45,6 +45,122 @@ static const int MCD_SIZE	= 1024 *  8  * 16;		// Legacy PSX card default size
...
 
+// ECC code ported from mymc
+// https://sourceforge.net/p/mymc-opl/code/ci/master/tree/ps2mc_ecc.py
+// Public domain license
+
+u32 CalculateECC(u8* buf)
+{
+	static bool tables_generated;
+	static u8 parity_table[256];
+	static u8 column_parity_mask[256];
```
Comment (ssakash) [inline pcsx2/gui/MemoryCardFile.cpp]: I know these are default zero initialized, but coverity is picky about it.

**Comment by lightningterror** [comment]
Is this ready? Was it tested if it works properly?

**Comment by KrossX** [comment]
It is not ready, and I would've liked some tests but testers will take their time. Check back in a few months or so.

**Comment by jLynx** [comment]
Are you still having checksum issues with the .bin files?

**Comment by KrossX** [comment]
Checksum issues? OPL VMCs can already be used on this branch. It just needs testing.

**Comment by jLynx** [comment]
Oh sorry, my mistake. I though you said in another post you were having checksum issues with it. but looking back I cant find it so I must have imagined it haha

**Comment by KrossX** [comment]
There are checksum issues with the format but as the PR description says, the extra bytes are generated on load and removed on save. OPL VMUs already work on my tests, but I need tests from users of such setups.

**Comment by jLynx** [comment]
I just tested with a few savegames and it seems to run the .bin files perfectly!

File: pcsx2/gui/MemoryCardFile.cpp [hunk 2]
```diff
@@ -45,6 +45,122 @@ static const int MCD_SIZE	= 1024 *  8  * 16;		// Legacy PSX card default size
...
+	wxFFile fin(file_in, "rb");
+
+	if (fin.IsOpened()) {
+		wxFFile fout(file_out, "wb");
+
+		if (fout.IsOpened()) {
+			size_t size = fin.Length();
+			u8* buffer = new u8[size];
+
+			if (buffer) {
```
[the remark this answers is not in the document]
Comment (KrossX) [reply pcsx2/gui/MemoryCardFile.cpp:105]: How about now? No need to worry about fancy `new`.

File: pcsx2/gui/MemoryCardFile.cpp [hunk 3]
```diff
@@ -45,6 +45,122 @@ static const int MCD_SIZE	= 1024 *  8  * 16;		// Legacy PSX card default size
...
+				for (size_t i = 0; i < (size / 512); i++) {
+					u8* buf = &buffer[i * 512];
+					fout.Write(buf, 512);
+
+					for (int j = 0; j < 4; j++) {
+						u32 checksum = CalculateECC(&buf[j * 128]);
+						u8 * cb = (u8*)&checksum;
+						fout.Write(cb++, 1);
+						fout.Write(cb++, 1);
+						fout.Write(cb, 1);
```
[the remark this answers is not in the document]
Comment (KrossX) [reply pcsx2/gui/MemoryCardFile.cpp]: First suggestion worked.

File: pcsx2/gui/Panels/MemoryCardListPanel.cpp [hunk 4]
```diff
@@ -1056,6 +1058,7 @@ void Panels::MemoryCardListPanel_Simple::ReadFilesAtMcdFolder(){
 	if ( memcardDir.IsOpened() ) {
 		// add memory card files
 		wxDir::GetAllFiles( filename, &memcardList, L"*.ps2", wxDIR_FILES );
+		wxDir::GetAllFiles( filename, &memcardList, L"*.bin", wxDIR_FILES);
```
Comment (refractionpcsx2) [inline pcsx2/gui/Panels/MemoryCardListPanel.cpp:1061]: can you put a space after wx_DIR_FILES?

File: pcsx2/gui/MemoryCardFile.cpp [hunk 5]
```diff
@@ -206,6 +314,18 @@ void FileMemoryCard::Open()
 		NTFS_CompressFile( str, g_Conf->McdCompressNTFS );
 #endif
 
+		if (str.EndsWith(".bin"))
+		{
+			wxString newname = str + "x";
+			if(!ConvertNoECCtoRAW(str, newname))
+			{
```
Comment (refractionpcsx2) [inline pcsx2/gui/MemoryCardFile.cpp:321]: put the bracket on the same line as the if to match the rest of the file, also space between if and (

**Review by refractionpcsx2 — CHANGES_REQUESTED** [review]

**Comment by lightningterror** [comment]
Or just run clang on the entire file, note you can set clang to ignore the arrays.

**Comment by refractionpcsx2** [comment]
Or clang them.

**Comment by KrossX** [comment]
Clang-format seems to have made, quite a lot of changes.

**Comment by refractionpcsx2** [comment]
It looks great, except the massacre it did to the parity tables xD

**Comment by KrossX** [comment]
How about now?

**Comment by refractionpcsx2** [comment]
That's better on the eye yea :)

**Pull request merged**